mlflow/utils/rest_utils.py

```diff
@@ -14,27 +12,91 @@
 from mlflow.utils.string_utils import strip_suffix
 from mlflow.exceptions import MlflowException, RestException
 
-_REST_API_PATH_PREFIX = "/api/2.0"
 RESOURCE_DOES_NOT_EXIST = "RESOURCE_DOES_NOT_EXIST"
-
-_logger = logging.getLogger(__name__)
-
+_REST_API_PATH_PREFIX = "/api/2.0"
 _DEFAULT_HEADERS = {"User-Agent": "mlflow-python-client/%s" % __version__}
+# Response codes that generally indicate transient network failures and merit client retries,
+# based on guidance from cloud service providers
+# (https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines)
+_TRANSIENT_FAILURE_RESPONSE_CODES = frozenset(
+    [
+        408,  # Request Timeout
+        429,  # Too Many Requests
+        500,  # Internal Server Error
+        502,  # Bad Gateway
+        503,  # Service Unavailable
+        504,  # Gateway Timeout
+    ]
+)
+
+
+def _get_http_response_with_retries(
+    method, url, max_retries, backoff_factor, retry_codes, **kwargs
+):
+    """
+    Performs an HTTP request using Python's `requests` module with an automatic retry policy.
+
+    :param method: a string indicating the method to use, e.g. "GET", "POST", "PUT".
+    :param url: the target URL address for the HTTP request.
+    :param max_retries: Maximum total number of retries.
+    :param backoff_factor: a time factor for exponential backoff. e.g. value 5 means the HTTP
+      request will be retried with interval 5, 10, 20... seconds. A value of 0 turns off the
+      exponential backoff.
+    :param retry_codes: a list of HTTP response error codes that qualifies for retry.
+    :param kwargs: Additional keyword arguments to pass to `requests.Session.request()`
+
+    :return requests.Response object.
+    """
+    assert 0 <= max_retries < 10
+    assert 0 <= backoff_factor < 120
+    retry = Retry(
+        total=max_retries,
+        connect=max_retries,
+        read=max_retries,
+        redirect=max_retries,
+        status=max_retries,
+        status_forcelist=retry_codes,
+        method_whitelist=None,  # Allow retry on any HTTP methods.
```

Comment:
```suggestion
        allowed_methods=None,  # Allow retry on any HTTP methods.
```

Can we use `allowed_methods` instead of `method_whitelist` that is deprecated now?

https://github.com/urllib3/urllib3/blob/main/CHANGES.rst#1260-2020-11-10

Reply:
On the master branch of the `url` repository, `method_whitelist` has already been removed:

https://github.com/urllib3/urllib3/blob/25ed51f319a3c6920bb5bd959924ca80925ee120/src/urllib3/util/retry.py#L188-L206

Reply:
I tried allowed_methods on my Python 3.6 setup, it didn't work. But we can try to see if passes integration tests.